### ISSUE TITLE
Support for Anchor Links

### ISSUE THREAD

**Issue by kylelaker** [EasyDynamics/oscal-react-library#176]
As a user of the OSCAL Viewer/Editor, I want to be able to quickly navigate to sections of OSCAL documents, including specific controls and control implementations, using anchor links so that I don't have to scroll through often large pages and can bookmark/share specific sections.

### Implementation Notes (and Original Submission)
Within the Catalog Viewer, it would be really helpful if I could link directly to a control within the catalog; for example, if navigating to https://oscal-viewer.msd.easydynamics.com/catalog?url=https://raw.githubusercontent.com/usnistgov/oscal-content/master/nist.gov/SP800-53/rev5/json/NIST_SP-800-53_rev5_catalog.json#SI-3 were to automatically result in "System and Information Integrity" expanded with "SI-3 Malicious Code Protection" focused in the viewport. Perhaps each control could also have a way to copy a link to it (for example, the way that you can do so for headings on the [Microsoft Docs](https://docs.microsoft.com/en-us/windows/wsl/install-on-server#extract-and-install-a-linux-distribution)).

The link should use the `#`/fragment in the URL rather than a querystring parameter for determining where in the page to navigate to, in order to support browsers' native capabilities to do so. This could either be handled by adding an `id` attribute to the headings or placing an `<a>` tag with a name directly above the `<hX>` tags.

I am not sure whether this feature request cleanly maps to the other viewer types.

**Comment by rgauss** [comment]
We'll likely want to hold off on this until after #88, but should think about how we'll address this during implementation there.

**Comment by rgauss** [comment]
We have a placeholder issue for more generic anchor links across the app, not just catalogs.

I'm going to broaden the scope of this issue to support that.

**Comment by kylelaker** [comment]
We need to make sure that we have something to track _creating_ links; either making a new ticket or ensuring it's supported as part of markup support or another ticket.

**Comment by Bronstrom** [comment]
Ideally we will have #88 finished before this issue is completed, however I believe most of the implementation can be done without having the issue completed.

When a hash is provided in the browser's url, it can be acquired, validated, and if there is a matching element to this hash on the page it will be navigated to. (As mentioned in the implementation notes, an `id` can be set to a header which matches this hash.) This functionality can be handled within a hook in `OSCALLoader`.

To acquire a heading like [Microsoft Docs](https://docs.microsoft.com/en-us/windows/wsl/install-on-server#extract-and-install-a-linux-distribution), a header or a [newly added icon](https://mui.com/material-ui/material-icons/?query=link&selected=Link) can be linked to a hash by surrounding it in a `react-router-dom` `Link`.

I was curious to what extent would we like to have anchor links supported? Would we want to link to the major sections of a page like `control-groups`, `system-characteristics`, and `back-matter`? Would we want the user to link to a specific resource in backmatter (like the title of a diagram) or a `party` in the Parties table? Or even as specific as a user within a `system-implementation/users` table?

**Comment by tuckerzp** [comment]
>  what extent would we like to have anchor links supported?

Here are my thoughts on what should be linked to

### Anchor links will be added at

#### Shared Sections
- [ ] Parties `parties`
- [ ] Profiles/Catalog Inheritance `profile-catalog-inheritance`
- [ ] Back Matter `back-matter`
  - [ ] Resources `resources`

#### Catalog Viewer
- [ ] Controls Groups `control-groups`
  - [ ] Control & Sub Controls `?url={url}#{control-id}

#### SSP Viewer
- [ ] System Characteristics `system-characteristics`
  - [ ] Authorization Boundary `authorization-boundary`
  - [ ] Network Architecture `network-architecture`
  - [ ] Data Flow `data-flow`
- [ ] System Implementation `system-implementation` 
  - [ ] Users `users`
  - [ ] Components `components`
  - [ ] Inventory Items `inventory-items`
- [ ] Control Implementation `control-implementation`

#### Component Viewer
- [ ] Components `components`
- [ ] Control Implementations `control-implementations`

#### Profile Viewer
- [ ] Imported Controls `mported-controls`
  - [ ] Control & Sub Controls `?url={url}#{control-id}

**Comment by Bronstrom** [comment]
3/10/23 end of Sprint 42 status: This is probably somewhere from 75%-90% complete. The functionality and design is practically finished, with a minor adjustment or two. The code is still in review and to be revisited - including some linting errors needing to be ironed out. The Cypress tests for this issue are also finished - but require for the code to be merged in oscal-react library in order for the tests to pass in the respective oscal-editor-deployment issue.

**Comment by Bronstrom** [comment]
3/10/23 Sprint Planning - Moved from Sprint 42 to Sprint 43

**Comment by Bronstrom** [comment]
3/23/23 end of Sprint 43 status: This issue is about 90% complete overall. I didn't get to dedicate as much of my attention to this issue during this sprint at I had hoped. Improvements have been made to anchor links and there were some refactoring choices made. Acquiring the fragment in the URL has been changed to be acquired in a more react like method - although this currently causes many tests to fail and will need to be decided on if this is the correct method. `control-groups` remains as the main issue holding back this issue from completion. There are still a couple of architecture decisions that will need to be made here - but I do believe this is getting closer to finishing.

**Comment by kylelaker** [comment]
Reopening this until the tests are in and we clean up final items mentioned on the PR. https://github.com/EasyDynamics/oscal-react-library/pull/600#pullrequestreview-1375646312

**Comment by Bronstrom** [comment]
4/7/23 Sprint Planning - Moved from Sprint 44 to Sprint 45
